File: plan2/week4/main.cpp
class Solution
{
public:
    int getMax(vector<int> &piles)
    {

        int maxNum = -1e6;

        for (int n : piles)
        {
            maxNum = max(n, maxNum);
        }

        return maxNum;
    }

    bool isValid(int h, int speed, vector<int> &piles)
    {
        int hours = 0;
        for (int i = 0; i < piles.size(); i++)
        {
            hours += (piles[i] + speed - 1) / speed;
        }
        return hours <= h;
    }

    int minEatingSpeed(vector<int> &piles, int h)
    {
        long long min = 1;
        long long max = getMax(piles);

        long long res = max;
        while (min <= max)
        {
            long long mid = min + (max - min) / 2;

            if (isValid(h, mid, piles))
            {
                res = mid;
                max = mid - 1;
            }
            else
            {
                min = mid + 1;
            }
        }

        return res;
    }
};

File: plan2/week4/go/main.go
package main

import (
	"math"
)

func getMax(piles []int) int {
	var maxCount = math.MinInt
	for i := 0; i < len(piles); i++ {
		if maxCount < piles[i] {
			maxCount = piles[i]
		}
	}

	return maxCount
}

func isValid(k int, piles []int, hours int) bool {
	var h = 0

	for i := 0; i < len(piles); i++ {
		h += (piles[i] + k - 1) / k
	}
	return h <= hours
}

func minEatingSpeed(piles []int, h int) int {

	var min = 1
	var max = getMax(piles)

	res := -1
	for min <= max {
		var mid = min + (max-min)/2

		if isValid(mid, piles, h) {
			res = mid
			max = mid - 1
		} else {
			min = mid + 1
		}
	}

	return res
}
